server: TestEnsureSQLStatsAreFlushedDuringDrain failed

server.TestEnsureSQLStatsAreFlushedDuringDrain [failed](https://mesolite.cluster.engflow.com/invocations/default/f04bad37-87e9-4851-9e49-625923b61cf4?testReportRun=3&testReportShard=10&testReportAttempt=1#targets-Ly9wa2cvc2VydmVyOnNlcnZlcl90ZXN0) on master @ [f43db402f520dd947d2aa3f742879ba2fe03e048](https://github.com/cockroachdb/cockroach/commits/f43db402f520dd947d2aa3f742879ba2fe03e048):

Fatal error:

```
panic: descriptor leak was detected for: 1 ({[%!s(*lease.descriptorVersionState=&{0xc00747c3c0 0xc008940300 {{} 1} {[] {} 0xc0132f12a0} {[] {} <nil>} {{{{0 0}}} 0xc0073dfb60}})]}) [recovered]
	panic: descriptor leak was detected for: 1 ({[%!s(*lease.descriptorVersionState=&{0xc00747c3c0 0xc008940300 {{} 1} {[] {} 0xc0132f12a0} {[] {} <nil>} {{{{0 0}}} 0xc0073dfb60}})]})

```
Stack: 

```
goroutine 6159807 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc00f8fcff0?, {0x7392a38, 0xc0180a7980})
	pkg/util/stop/stopper.go:226 +0x65
panic({0x604ce20?, 0xc01d0ec890?})
	GOROOT/src/runtime/panic.go:791 +0x132
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).SetDraining.func1(0xc00747c3c0, {0x7392a00, 0xc0035ee320}, 0x1)
	pkg/sql/catalog/lease/lease.go:1601 +0x1db
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).SetDraining(0xc001f22008, {0x7392a00, 0xc0035ee320}, 0x1, 0xc00f8fd448, 0x1)
	pkg/sql/catalog/lease/lease.go:1604 +0x187
github.com/cockroachdb/cockroach/pkg/server.(*drainServer).drainClients(0xc00cff9450, {0x7395378?, 0xc00e3b2aa0?}, 0xc0070d5448)
	pkg/server/drain.go:489 +0x4ce
github.com/cockroachdb/cockroach/pkg/server.(*drainServer).drainInner(0xc00cff9450, {0x7395378, 0xc00e3b2aa0}, 0xc00f8fd448, 0x0)
	pkg/server/drain.go:383 +0x165
github.com/cockroachdb/cockroach/pkg/server.(*drainServer).runDrain(0xc00cff9450, {0x7395378, 0xc00e3b2aa0}, 0x0)
	pkg/server/drain.go:343 +0x15c
github.com/cockroachdb/cockroach/pkg/server.(*drainServer).handleDrain(0xc00cff9450, {0x7395378, 0xc00e3b2aa0}, 0xc00b088cc0, {0x7fee472a5f08, 0xc0148ba130})
	pkg/server/drain.go:193 +0x128
github.com/cockroachdb/cockroach/pkg/server.(*adminServer).drain(0xc002f70d00, 0xc00b088cc0, {0x7fee472a5f08, 0xc0148ba130})
	pkg/server/drain.go:120 +0x132
github.com/cockroachdb/cockroach/pkg/server.(*adminServer).Drain(...)
	pkg/server/drain.go:91
github.com/cockroachdb/cockroach/pkg/server/serverpb._Admin_Drain_Handler({0x65c7080, 0xc00e225740}, {0x73c8118, 0xc0036784b0})
	bazel-out/k8-fastbuild/bin/pkg/server/serverpb/serverpb_go_proto_/github.com/cockroachdb/cockroach/pkg/server/serverpb/admin.pb.go:6180 +0x107
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.StreamServerInterceptor.func13({0x65c7080, 0xc00e225740}, {0x73c8118, 0xc0036784b0}, 0xc01e30b3b0, 0x6a01700)
	pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:123 +0x48e
google.golang.org/grpc.getChainStreamHandler.func1({0x65c7080, 0xc00e225740}, {0x73c8118, 0xc0036784b0})
	external/org_golang_google_grpc/server.go:1504 +0xb2
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func4({0x65c7080, 0xc00e225740}, {0x73c8118, 0xc0036784b0}, 0xc01e30b3b0?, 0xc00f7daf00)
	pkg/rpc/context.go:188 +0x6c
google.golang.org/grpc.getChainStreamHandler.func1({0x65c7080, 0xc00e225740}, {0x73c8118, 0xc0036784b0})
	external/org_golang_google_grpc/server.go:1504 +0xb2
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func2.1({0xc00252e750?, 0x7392a38?})
	pkg/rpc/context.go:146 +0x27
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0xc00252e750, {0x7392a38, 0xc0180a7980}, {0xc01c6ca060, 0x26}, 0xc0031b8ab0)
	pkg/util/stop/stopper.go:349 +0x13c
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func2({0x65c7080, 0xc00e225740}, {0x73c8118, 0xc0036784b0}, 0xc01e30b3b0, 0xc00f7dae80)
	pkg/rpc/context.go:145 +0xcd
google.golang.org/grpc.NewServer.chainStreamServerInterceptors.chainStreamInterceptors.func2({0x65c7080, 0xc00e225740}, {0x73c8118, 0xc0036784b0}, 0xc01e30b3b0, 0xc0148ba120?)
	external/org_golang_google_grpc/server.go:1495 +0x85
google.golang.org/grpc.(*Server).processStreamingRPC(0xc01bdb2000, {0x73e8300, 0xc00b422d00}, 0xc00cca67e0, 0xc00b15e9c0, 0xb0ebb40, 0x0)
	external/org_golang_google_grpc/server.go:1659 +0x1285
google.golang.org/grpc.(*Server).handleStream(0xc01bdb2000, {0x73e8300, 0xc00b422d00}, 0xc00cca67e0, 0x0)
	external/org_golang_google_grpc/server.go:1739 +0x999
google.golang.org/grpc.(*Server).serveStreams.func1.1()
	external/org_golang_google_grpc/server.go:970 +0xaf
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 6112423
	external/org_golang_google_grpc/server.go:981 +0x136
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestEnsureSQLStatsAreFlushedDuringDrain
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestEnsureSQLStatsAreFlushedDuringDrain1767872514
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:143: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=3</code>
 - <code>shard=10</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestEnsureSQLStatsAreFlushedDuringDrain.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-52831

--- a/pkg/sql/catalog/lease/lease.go
+++ b/pkg/sql/catalog/lease/lease.go
@@ -64,6 +64,7 @@ import (
 
 var errRenewLease = errors.New("renew lease on id")
 var errReadOlderVersion = errors.New("read older descriptor version from store")
+var errLeaseManagerIsDraining = errors.New("cannot acquire lease when draining")
 
 // LeaseDuration controls the duration of sql descriptor leases.
 var LeaseDuration = settings.RegisterDurationSetting(
@@ -882,7 +883,7 @@ func acquireNodeLease(
 		},
 		func(ctx context.Context) (interface{}, error) {
 			if m.IsDraining() {
-				return nil, errors.New("cannot acquire lease when draining")
+				return nil, errLeaseManagerIsDraining
 			}
 			newest := m.findNewest(id)
 			var currentVersion descpb.DescriptorVersion
@@ -1359,6 +1360,9 @@ func (m *Manager) AcquireByName(
 	parentSchemaID descpb.ID,
 	name string,
 ) (LeasedDescriptor, error) {
+	if m.IsDraining() {
+		return nil, errLeaseManagerIsDraining
+	}
 	// When offline descriptor leases were not allowed to be cached,
 	// attempt to acquire a lease on them would generate a descriptor
 	// offline error. Recent changes allow offline descriptor leases
@@ -1531,6 +1535,9 @@ func (m *Manager) Acquire(
 	ctx context.Context, timestamp hlc.Timestamp, id descpb.ID,
 ) (LeasedDescriptor, error) {
 	for {
+		if m.IsDraining() {
+			return nil, errLeaseManagerIsDraining
+		}
 		t := m.findDescriptorState(id, true /*create*/)
 		desc, _, err := t.findForTimestamp(ctx, timestamp)
 		if err == nil {
@@ -1602,8 +1609,12 @@ func (m *Manager) SetDraining(
 			t.mu.Lock()
 			defer t.mu.Unlock()
 			leasesToRelease := t.removeInactiveVersions(ctx)
-			// Ensure that all leases are released at this time.
-			if buildutil.CrdbTestBuild && assertOnLeakedDescriptor && len(t.mu.active.data) > 0 {
+			// Ensure that all leases are released at this time. Ignore any descriptors
+			// that have acquisitions in progress.
+			if buildutil.CrdbTestBuild &&
+				assertOnLeakedDescriptor &&
+				len(t.mu.active.data) > 0 &&
+				t.mu.acquisitionsInProgress == 0 {
 				// Panic that a descriptor may have leaked.
 				panic(errors.AssertionFailedf("descriptor leak was detected for: %d (%s)", t.id, t.mu.active))
 			}

